src/lib.rs

```diff
@@ -1,20 +1,124 @@
 mod error;
-mod graph;
-mod loader;
-mod parser;
-mod reqwest_loader;
-mod resolve_import;
-
-pub use error::Error;
-pub use graph::ModuleGraph;
-pub use graph::GRAPH_VERSION;
-pub use loader::MemoryLoader;
-pub use loader::ModuleInfo;
-pub use loader::ModuleLoad;
-pub use loader::ModuleLoadFuture;
-pub use loader::ModuleLoader;
-pub use loader::ModuleSource;
-pub use loader::ModuleStream;
-pub use reqwest_loader::load_reqwest;
-pub use reqwest_loader::none_middleware;
-pub use reqwest_loader::ReqwestLoader;
+pub mod v1;
+pub mod v2;
+
+use std::pin::Pin;
+use std::sync::Arc;
+
+use futures::Future;
+use serde::Deserialize;
+use serde::Serialize;
+use tokio::io::AsyncBufReadExt;
+use tokio::io::AsyncReadExt;
+use v2::ESZIP_V2_MAGIC;
+
+pub use crate::error::ParseError;
+pub use crate::v1::EsZipV1;
+pub use crate::v2::EsZipV2;
+
+pub use deno_ast;
+pub use deno_graph;
+
+pub enum EsZip {
+  V1(EsZipV1),
+  V2(EsZipV2),
+}
+
+/// This future needs to polled to parse the eszip file.
+type EszipParserFuture<R> =
+  Pin<Box<dyn Future<Output = Result<tokio::io::BufReader<R>, ParseError>>>>;
+
+impl EsZip {
+  /// Parse a byte stream into an EsZip. This function completes when the header
+  /// is fully received. This does not mean that the entire file is fully
+  /// received or parsed yet. To finish parsing, the future returned by this
+  /// function in the second tuple slot needs to be polled.
+  pub async fn parse<R: tokio::io::AsyncRead + Unpin + 'static>(
+    reader: R,
+  ) -> Result<(EsZip, EszipParserFuture<R>), ParseError> {
+    let mut reader = tokio::io::BufReader::new(reader);
+    reader.fill_buf().await?;
+    let buffer = reader.buffer();
+    if buffer.len() >= 8 && &buffer[..8] == ESZIP_V2_MAGIC {
```

Comment:
If I'm parsing a V2 eszip and the header magic has not completely arrived yet, this check will fail and it will be parsed as an eszip v1?

Reply:
The `reader.fill_buf()` will do one `read` syscall. There is no way the first packet contains less than the first 8 bytes of the data. If it does, we have problems elsewhere.